src/history/hs.c

```diff
@@ -823,6 +861,28 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
 
         if (modifies.size > 0)
             WT_STAT_CONN_INCR(session, cache_hs_write_squash);
+
+        /*
+         * We need to clear the history store if we haven't inserted anything into the history store
+         * and there are updates without timestamps in the middle of the update chain.
+         *
+         * e.g., U@10 -> T@0 -> U@5.
+         *
+         * But we don't need to clear the history store if we write an update without timestamp to
+         * the data store because we don't insert any update with timestamp to the history store and
+         * we will clear the history store again once that update is moved to the history store.
+         *
+         * e.g., U@0 -> U@10 -> U@5 and U@1 in the history store. U@10 and U@5 are not inserted to
+         * the history store as they are flagged as WT_UPDATE_MASKED_BY_NON_TS_UPDATE and U@1 is not
+         * removed from the history store. U@1 will be removed from the history store once U@0 is
+         * moved to the history store.
+         */
+        if (clear_hs && (first_non_ts_upd->txnid != list->onpage_upd->txnid ||
+                          first_non_ts_upd->start_ts != list->onpage_upd->start_ts)) {
+            /* We can only delete history store entries that have timestamps. */
+            WT_ERR(__wt_hs_delete_key_from_ts(session, btree->id, key, 1));
```

Comment:
Is it fine to delete down here? I feel like we may remove timestamped content that we just inserted.

If you had an update chain like:
```
U@10 (new onpage value) => U@9 => U@0 (first_non_ts_upd) => some other updates already in history store
```

I think `U@9` would go missing.

Reply:
With the change, we will only delete hs content once for each chain. In your example, when we insert U@0, we will delete hs content and set clear_hs to false. So we will not delete hs content again outside the loop.

Reply:
Ah, understood. Yeah this making more sense now.